flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/feature/extraction/FeatureHasher.scala

```diff
@@ -0,0 +1,142 @@
+  *            val transformer = FeatureHasher().setNumFeatures(65536).setNonNegative(false)
+  *
+  *            transformer.transform(trainingDS)
+  *          }}}
+  *
+  * =Parameters=
+  *
+  * - [[FeatureHasher.NumFeatures]]: The number of features (entries) in the output vector;
+  * by default equal to 2&#94;20
+  * - [[FeatureHasher.NonNegative]]: Whether output vector should contain non-negative values only.
+  * When True, output values can be interpreted as frequencies. When False, output values will have
+  * expected value zero; by default equal to false
+  */
+class FeatureHasher extends Transformer[Seq[String], Vector] with Serializable {
+
+  // The seed used to initialize the hasher
+  val Seed = 0
+
+  /** Sets the number of features (entries) in the output vector
+    *
+    * @param numFeatures the user-specified numFeatures value. In case the user gives a value less
+    *                    than 1, numFeatures is set to its default value: 2&#94;20
+    * @return the FeatureHasher instance with its numFeatures value set to the user-specified value
+    */
+  def setNumFeatures(numFeatures: Int): FeatureHasher = {
+    // number of features must be greater zero
+    if(numFeatures < 1) {
+      return this
+    }
+    parameters.add(NumFeatures, numFeatures)
+    this
+  }
+
+  /** Sets whether output vector should contain non-negative values only
+    *
+    * @param nonNegative the user-specified nonNegative value.
+    * @return the FeatureHasher instance with its nonNegative value set to the user-specified value
+    */
+  def setNonNegative(nonNegative: Boolean): FeatureHasher = {
+    parameters.add(NonNegative, nonNegative)
+    this
+  }
+
+  override def transform(input: DataSet[Seq[String]], parameters: ParameterMap):
+  DataSet[Vector] = {
+    val resultingParameters = this.parameters ++ parameters
+
+    val nonNegative = resultingParameters(NonNegative)
+    val numFeatures = resultingParameters(NumFeatures)
+
+    // each item of the sequence is hashed and transformed into a tuple (index, value)
+    input.map {
+      inputSeq => {
+        val entries = inputSeq.map {
+          s => {
+            // unicode strings are converted to utf-8
+            // bytesHash is faster than arrayHash, because it hashes 4 bytes at once
+            val h = MurmurHash3.bytesHash(s.getBytes(Charset.forName("UTF-8")), Seed) % numFeatures
+            val index = Math.abs(h)
+            /* instead of using two hash functions (Weinberger et al.), assume the sign is in-
+               dependent of the other bits */
+            val value = if(h >= 0) 1.0 else -1.0
+            (index, value)
+          }
+        }
+        val myVector = SparseVector.fromCOO(numFeatures, entries)
+        // in case of non negative output, return the absolute of the vector
+        if(nonNegative) {
+          // maybe there's a straightforward way of applying a function element wise to a vector
```

Comment:
It's OK, it desugars to 
```scala
myVector.indices.foreach(...)
```
which is as precise as it gets. In general, using comprehensions makes for more readable code.